### DISCUSSION TITLE
set directory to store the compiler artifacts

### DISCUSSION THREAD

**Discussion by Koushikphy** [fortran-lang/fortls#149] [Q&A] [answered]
I was wondering if there is a way to set the object directory for `fortls`. Currently when I'm using `forls` (with Modern Fortran) in vscode it creates the `.mod` files directly in the main source directory. Can I specify a different directory to store and use those `.mod` files for `fortls` to use so that it don't pollute my source directory.

**Comment by gnikit** [comment]
`fortls` does not use the compiler, it is platform agnostic. That is Modern Fortran's linter (gfortran by default). You can control the output of your .mod files by setting the option `  "fortran.linter.modOutput": "your-path-here"`. Make sure however that you also include this path to `  "fortran.linter.includePaths": []` otherwise your linter will not work.

**Reply by gnikit** [reply]
What `fortran.linter.modOutput` does is basically add the `-J` flag when calling gfortran

**Comment by Koushikphy** [comment]
I'm using `ifort` for linting and did `"fortran.linter.modOutput"`, it did not work. Also tried adding `"-module ../obj -I ../obj"` to linter extra args didn't work.

**Comment by gnikit** [comment]
Can you post the logs from the log channel? I can't tell much from this. Also, the paths get resolved against your project's root, `../obj` would be outside that. Does that directory exist?

**Comment by Koushikphy** [comment]
I have this as my settings
```json
    "fortran.linter.modOutput":"../obj",
    "fortran.linter.includePaths": ["../obj"],
    "fortran.linter.compiler": "ifort",
```
I tried this with `gfortran` it works properly. Thats where I noticed I have to put `../obj` that is outside of `src` directory in the project root. Also how do I get the log?

**Reply by gnikit** [reply]
Can you post the Output of the Modern Fortran extension e.g.
![image](https://user-images.githubusercontent.com/16143716/174453013-832fc30d-a4f3-4b0e-91b1-850209ada669.png)

**Comment by Koushikphy** [comment]
Here's the output for `ifort`
```bash
["INFO" - 12:48:47 am] using linter: ifort located in: /opt/intel/oneapi/compiler/2021.3.0/linux/bin/intel64/ifort
["INFO" - 12:48:47 am] Linter.arguments:
-warn
all
["INFO" - 12:48:47 am] Linter.moduleOutput: -module ../obj
["INFO" - 12:48:47 am] Initialising Language Server for workspace: /home/koushik/TMP/111/src/test.f90 with command-line options: --enable_code_actions, --hover_signature, --use_signature_help, --lowercase_intrinsics, --nthreads=4, --incremental_sync
[INFO - 00:48:47] fortls - Fortran Language Server 2.9.0 Initialized
[ERRO - 00:48:47] error handling request: {'jsonrpc': '2.0', 'method': 'workspace/didChangeConfiguration', 'params': {'settings': {'fortran': {'provide': {'autocomplete': 'fortls', 'hover': 'fortls', 'symbols': 'fortls'}, 'preferredCase': 'lowercase', 'linter': {'compiler': 'ifort', 'includePaths': ['../obj'], 'compilerPath': '', 'extraArgs': [], 'modOutput': '../obj'}, 'formatting': {'formatter': 'fprettify', 'findentArgs': [], 'fprettifyArgs': ['-i 4', '-w 2', '--enable-replacements', '--c-relations', '--enable-decl', '--whitespace-intrinsics=False', '--whitespace-print=False'], 'path': ''}, 'fortls': {'path': 'fortls', 'configure': '', 'notifyInit': False, 'incrementalSync': True, 'nthreads': 4, 'sortKeywords': False, 'symbolTypes': True, 'suffixes': [], 'directories': [], 'excludeSuffixes': [], 'excludeDirectories': [], 'preprocessor': {'suffixes': [], 'directories': [], 'definitions': {}}, 'disableDiagnostics': False, 'maxLineLength': -1, 'maxCommentLineLength': -1, 'extraArgs': [], 'disableAutoupdate': False, 'disabled': False, 'preserveKeywordOrder': None}, 'includePaths': None, 'gfortranExecutable': None, 'linterEnabled': None, 'linterExtraArgs': None, 'linterModOutput': None, 'provideSymbols': None, 'symbols': None, 'provideHover': None, 'provideCompletion': None}}}}
Traceback (most recent call last):
  File "/home/koushik/.local/lib/python3.8/site-packages/fortls/langserver.py", line 160, in handle
    handler(request)
  File "/home/koushik/.local/lib/python3.8/site-packages/fortls/langserver.py", line 1487, in serve_default
    raise JSONRPC2Error(
fortls.langserver.JSONRPC2Error
```

**Reply by gnikit** [reply]
~~You are right I spotted a bug, the option which we fetch from the VS Code settings has the wrong name. I will fix it now.~~

EDIT: I was wrong

**Comment by gnikit** [comment]
So I think I got to the bottom of this. The Intel compilers when using the syntax and semantics checking option `-syntax-only` do not create any sort of objects. So if you are using multiple modules you would have to open every single one of them to produce a `.mod` at the modOutput location which can then be included, **OR** just compile your code before and have the directories containing the `.mod` files being included in your linter include paths.

As far as I can tell this is a limitation from `ifort` since `gfortran` will recursively produce semantics for all `use`d modules

**Comment by Koushikphy** [comment]
But if I just manually open every single file, the mod files are created but in the src directory not in the object directory.

**Comment by Koushikphy** [comment]
Also needed to mention in case of gfortran, I also have to manually open each file in vs code for the mod file to be created. They are not automatically created when the workspace is opened

**Comment by gnikit** [comment, marked as answer]
The original issue has now been fixed in https://github.com/fortran-lang/vscode-fortran-support/pull/539.
As for the .mod file generation I was mistaken, the linter in general will not produce .mod files for files that are not open.

**Comment by Koushikphy** [comment]
Okay I will check once it gets released.
